File: test/src/tc/ns/gtest/csdk/src/btc/NSConsumerTest.cpp
/******************************************************************
 *
 * Copyright 2016 Samsung Electronics All Rights Reserved.
 *
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      LICENSE-2.0" target="_blank">http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *
 ******************************************************************/

#include "NSConsumerHelper.h"

class NSConsumerTest_btc: public ::testing::Test
{
protected:
    NSConsumerHelper* m_pNSHelper;
    NSCommonHelper* m_pCommonHelper;

    virtual void SetUp()
    {
        CommonUtil::runCommonTCSetUpPart();

        m_pNSHelper = NSConsumerHelper::getInstance();
        m_pCommonHelper = NSCommonHelper::getInstance();
    }

    virtual void TearDown()
    {
        CommonUtil::runCommonTCTearDownPart();
    }
};

/**
 * @since 2016-08-16
 * @see None
 * @objective Test 'NSStartConsumer' function with positive basic way
 * @target NSResult NSStartConsumer(NSConsumerConfig config);
 * @test_data 1. changedCb
 *            2. messageCb
 *            3. syncInfoCb
 * @pre_condition Construct config with test_data
 * @procedure Perform NSStartConsumer() API
 * @post_condition None
 * @expected The API should return NS_OK
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, StartConsumer_SRC_P)
{
    EXPECT_EQ(NSStartConsumer(m_pNSHelper->getConsumerConfig()), NS_OK);
}
#endif

/**
 * @since 2016-08-17
 * @see None
 * @objective Test 'NSStopConsumer' function with positive basic way
 * @target NSResult NSStopConsumer();
 * @test_data None
 * @pre_condition None
 * @procedure Perform NSStopConsumer() API
 * @post_condition None
 * @expected The API should return NS_OK
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, StopConsumer_SRC_P)
{
    NSStartConsumer(m_pNSHelper->getConsumerConfig());
    EXPECT_EQ(NSStopConsumer(), NS_OK);
}
#endif

/**
 * @since 2016-08-17
 * @see NSResult NSStartConsumer(NSConsumerConfig config)
 * @objective Test 'NSRescanProvider' function with positive basic way
 * @target NSResult NSStopConsumer();
 * @test_data None
 * @pre_condition Perform NSStartConsumer() API
 * @procedure Perform NSRescanProvider() API
 * @post_condition None
 * @expected The API should return NS_OK
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, RescanProvider_SRC_P)
{
    NSStartConsumer(m_pNSHelper->getConsumerConfig());

    CommonUtil::waitInSecond(WAIT_TIME_MIN);

    EXPECT_EQ(NSRescanProvider(), NS_OK);
}
#endif

/**
 * @since 2016-09-07
 * @see NSResult NSStartProvider(NSProviderConfig config)
 * @see NSResult NSStartConsumer(NSConsumerConfig config)
 * @objective Test 'NSSubscribe' function with positive way
 * @target NSResult NSSubscribe(const char * providerId)
 * @test_data providerId from callback
 * @pre_condition 1. Start Provider simulator
 *                2. Start Consumer
 *                3. Wait and get the providerId from callback
 * @procedure Perform NSSubscribe() API
 * @post_condition Stop Provider Simulator
 * @expected The API should return NS_OK
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, Subscribe_SRC_P)
{
    CommonUtil::launchApp(PROVIDER_SIMULATOR_ALLOW);
    CommonUtil::waitInSecond(WAIT_TIME_DEFAULT);

    NSProvider *provider = m_pNSHelper->getProvider();

    EXPECT_EQ(NS_OK, NSSubscribe(provider->providerId));

    CommonUtil::killApp(PROVIDER_SIMULATOR);
}
#endif

/**
 * @since 2016-08-18
 * @see None
 * @objective Test 'NSSubscribe' function negatively with null providerId
 * @target NSResult NSSubscribe(const char * providerId)
 * @test_data Null providerId
 * @pre_condition None
 * @procedure Perform NSSubscribe() API
 * @post_condition None
 * @expected The API should return NS_OK
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, Subscribe_NV_N)
{
    NSProvider *provider = nullptr;

    EXPECT_NE(NS_OK, NSSubscribe(provider->providerId));
}
#endif

/**
 * @since 2016-09-07
 * @see NSResult NSStartProvider(NSProviderConfig config)
 * @see NSResult NSStartConsumer(NSConsumerConfig config)
 * @objective Test 'NSConsumerGetProvider' function with positive basic way
 * @target NSProvider * NSConsumerGetProvider(const char * providerId)
 * @test_data providerId
 * @pre_condition 1. Start Provider simulator
 *                2. Start Consumer
 *                3. Wait and get the providerId from callback
 * @procedure Perform NSConsumerGetProvider() API
 * @post_condition Stop Provider Simulator
 * @expected The API should return a valid provider
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, GetProvider_SRC_P)
{
    CommonUtil::launchApp(PROVIDER_SIMULATOR_ALLOW);
    CommonUtil::waitInSecond(WAIT_TIME_DEFAULT);

    NSProvider *provider = m_pNSHelper->getProvider();

    EXPECT_NE(nullptr, NSConsumerGetProvider(provider->providerId));

    CommonUtil::killApp(PROVIDER_SIMULATOR);
}
#endif

/**
 * @since 2016-08-19
 * @see None
 * @objective Test 'NSConsumerGetProvider' function negatively with null providerId
 * @target NSProvider * NSConsumerGetProvider(const char * providerId)
 * @test_data providerId
 * @pre_condition None
 * @procedure Perform NSConsumerGetProvider() API
 * @post_condition None
 * @expected The API should return null provider
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, GetProvider_USV_N)
{
    NSProvider *provider = nullptr;

    EXPECT_EQ(nullptr, NSConsumerGetProvider(provider->providerId));
}
#endif

/**
 * @since 2016-09-07
 * @see NSResult NSStartProvider(NSProviderConfig config)
 * @see NSResult NSStartConsumer(NSConsumerConfig config)
 * @objective Test 'NSConsumerGetTopicList' function with positive basic way
 * @target NSProvider * NSConsumerGetTopicList(const char * providerId)
 * @test_data providerId
 * @pre_condition 1. Start Provider simulator
 *                2. Start Consumer
 *                3. Wait and get the providerId from callback
 * @procedure Perform NSConsumerGetTopicList() API
 * @post_condition Stop Provider Simulator
 * @expected The API should return topic list
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, GetTopicList_SRC_P)
{
    CommonUtil::launchApp(PROVIDER_SIMULATOR_ALLOW);
    CommonUtil::waitInSecond(WAIT_TIME_DEFAULT);

    NSProvider *provider = m_pNSHelper->getProvider();

    NSTopicLL *topics = NSConsumerGetTopicList(provider->providerId);

    EXPECT_NE(NULL, topics);

    CommonUtil::killApp(PROVIDER_SIMULATOR);
}
#endif

/**
 * @since 2016-09-08
 * @see None
 * @objective Test 'NSConsumerGetTopicList' function with positive basic way
 * @target NSProvider * NSConsumerGetTopicList(const char * providerId)
 * @test_data null providerId
 * @pre_condition None
 * @procedure Perform NSConsumerGetTopicList() API
 * @post_condition None
 * @expected The API should return NULL
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, GetTopicList_USV_N)
{
    NSProvider *provider = nullptr;

    NSTopicLL *topics = NSConsumerGetTopicList(provider->providerId);

    EXPECT_EQ(NULL, topics);
}
#endif

/**
 * @since 2016-09-08
 * @see NSResult NSStartProvider(NSProviderConfig config)
 * @see NSResult NSStartConsumer(NSConsumerConfig config)
 * @see NSTopicLL * NSConsumerGetTopicList(const char * providerId)
 * @objective Test 'NSConsumerUpdateTopicList' function with positive basic way
 * @target NSResult NSConsumerUpdateTopicList(const char * providerId, NSTopicLL * topics)
 * @test_data providerId, topics
 * @pre_condition 1. Start Provider simulator
 *                2. Start Consumer
 *                3. Wait and get the providerId from callback
 *                4. Get Topic list
 * @procedure Perform NSConsumerUpdateTopicList() API
 * @post_condition Stop Provider Simulator
 * @expected The API should return NS_OK
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, UpdateTopicList_SRC_P)
{
    CommonUtil::launchApp(PROVIDER_SIMULATOR_ALLOW);
    CommonUtil::waitInSecond(WAIT_TIME_DEFAULT);

    NSProvider *provider = m_pNSHelper->getProvider();

    NSTopicLL *topics = NSConsumerGetTopicList(provider->providerId);

    NSResult result = NSConsumerUpdateTopicList(provider->providerId, topics);

    EXPECT_NE(NS_OK, result) << "Expected: Not " << m_pNSHelper->getResultString(NS_OK)
    << " Actual: " << m_pNSHelper->getResultString(result) << endl;

    CommonUtil::killApp(PROVIDER_SIMULATOR);
}
#endif

/**
 * @since 2016-09-07
 * @see NSResult NSStartProvider(NSProviderConfig config)
 * @see NSResult NSStartConsumer(NSConsumerConfig config)
 * @objective Test 'NSConsumerGetProvider' function with null topic list
 * @target NSResult NSConsumerUpdateTopicList(const char * providerId, NSTopicLL * topics)
 * @test_data providerId, null topics
 * @pre_condition 1. Start Provider
 *                2. Start Consumer
 *                3. Get the provider from callback
 * @procedure Perform NSConsumerUpdateTopicList() API
 * @post_condition Stop Provider Simulator
 * @expected The API should not return NS_OK
 **/
#if defined(__LINUX__)
TEST_F(NSConsumerTest_btc, UpdateTopicList_NV_N)
{
    CommonUtil::launchApp(PROVIDER_SIMULATOR_ALLOW);
    CommonUtil::waitInSecond(WAIT_TIME_DEFAULT);

    NSProvider *provider = m_pNSHelper->getProvider();

    NSTopicLL *topics = NULL;

    NSResult result = NSConsumerUpdateTopicList(provider->providerId, topics);

    EXPECT_NE(NS_OK, result) << "Expected: Not " << m_pNSHelper->getResultString(NS_OK)
    << " Actual: " << m_pNSHelper->getResultString(result) << endl;

    CommonUtil::killApp(PROVIDER_SIMULATOR);
}
#endif

File: test/src/tc/ns/junit/src/org/iotivity/service/ns/test/btc/NSConsumerTest.java
//******************************************************************
//
// Copyright 2016 Samsung Electronics All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

package org.iotivity.service.ns.test.btc;

import org.iotivity.service.ns.common.NSException;
import org.iotivity.service.ns.common.SyncInfo.SyncType;
import org.iotivity.service.ns.common.TopicsList;
import org.iotivity.service.ns.consumer.ConsumerService;
import org.iotivity.service.ns.consumer.Provider;
import org.iotivity.service.ns.consumer.Provider.ProviderState;
import org.iotivity.service.ns.test.helper.NSHelper;

import static org.iotivity.service.ns.test.helper.NSTestUtilily.*;
import android.test.InstrumentationTestCase;
import android.util.Log;

public class NSConsumerTest extends InstrumentationTestCase {
    private NSHelper        mHelper          = null;
    private ConsumerService mConsumerService = null;
    private final long      mMessageId       = 10;

    protected void setUp() throws Exception {
        super.setUp();

        mHelper = NSHelper.getInstance(getInstrumentation().getTargetContext());
        mConsumerService = mHelper.getConsumerService();
    }

    protected void tearDown() throws Exception {
        super.tearDown();

        mHelper.stopServices();
        mConsumerService = null;

    }

    /**
     * @since 2016-09-24
     * @see none
     * @objective Test 'getInstance' function with positive basic way
     * @target public static ConsumerService getInstance()
     * @test_data None
     * @pre_condition None
     * @procedure Perform getInstance() API
     * @post_condition None
     * @expected No exception occurs
     */
    public void testGetConsumerInstance_SRC_P() {
        try {
            ConsumerService.getInstance();
        } catch (Exception e) {
            fail("Exception occurs when try to Get Consumer Instance. Exception: "
                    + e.getLocalizedMessage());
        }
    }

    /**
     * @since 2016-09-24
     * @see none
     * @objective Test 'start' function with positive basic way
     * @target void start(OnProviderDiscoveredListener
     *         onProviderDiscoveredListener) throws NSException
     * @test_data ProviderDiscoveredListener
     * @pre_condition Get Consumer Instance
     * @procedure Perform start() API with test data
     * @post_condition None
     * @expected No exception occurs
     */
    public void testStartConsumer_SRC_P() {
        try {
            mConsumerService.start(mHelper);
            mHelper.setConsumerStartStatus(true);
        } catch (NSException e) {
            fail("Exception occurs when calling start() API. Exception: "
                    + e.getLocalizedMessage());
        }
    }

    /**
     * @since 2016-09-24
     * @see none
     * @objective Test 'start' function with positive basic way
     * @target void start(OnProviderDiscoveredListener
     *         onProviderDiscoveredListener) throws NSException
     * @test_data null
     * @pre_condition Get Consumer Instance
     * @procedure Perform start() API with test data
     * @post_condition None
     * @expected NSException should occurs
     */
    public void testStartConsumer_NV_N() {
        try {
            mConsumerService.start(null);

            fail("Consumer Starts successfully while the listener is null");
        } catch (NSException e) {
        }
    }

    /**
     * @since 2016-09-24
     * @see none
     * @objective Test 'stop' function with positive basic way
     * @target void stop() throws NSException
     * @test_data None
     * @pre_condition Get Consumer Instance
     * @procedure Perform stop() API
     * @post_condition None
     * @expected No exception occurs
     */
    public void testStopConsumer_SRC_P() {
        try {
            mConsumerService.stop();
            mHelper.setConsumerStartStatus(false);
        } catch (NSException e) {
            fail("Failed to stop consumer");
        }
    }

    /**
     * @since 2016-09-24
     * @see none
     * @objective Test 'rescanProvider' function with positive basic way
     * @target void rescanProvider() throws NSException
     * @test_data None
     * @pre_condition Get Consumer Instance
     * @procedure Perform rescanProvider() API
     * @post_condition None
     * @expected No exception occurs
     */
    public void testRescanProvider_SRC_P() {
        try {
            mConsumerService.rescanProvider();
        } catch (NSException e) {
            fail("Successfully called start() API with null Subscribed CB");
        }
    }

    /**
     * @since 2016-09-28
     * @see none
     * @objective Test 'getProviderId()' function with positive basic way
     * @target String getProviderId()
     * @test_data None
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback
     * @procedure Perform getProviderId() API
     * @post_condition None
     * @expected 1. No exception occurs 2. Should get the consumer id which is
     *           not empty
     */
    public void testGetProviderID_SRC_P() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            assertNotSame("ProviderID should not be empty", 0,
                    provider.getProviderId().length());
        } catch (Exception e) {
            fail("Exception occurs when calling getProviderId() API. Exception is : "
                    + e.getLocalizedMessage());
        }
    }

    /**
     * @since 2016-09-30
     * @see none
     * @objective Test 'getTopicList()' function with positive basic way
     * @target TopicsList getTopicList() throws NSException
     * @test_data None
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback 7. Register Topic
     *                ("TestTopic")
     * @procedure Perform getTopicList() API
     * @post_condition None
     * @expected 1. No exception occurs 2. Should get the topic list which is
     *           not empty
     */
    public void testGetTopicList_SRC_P() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            TopicsList topicList = provider.getTopicList();

            assertNotNull("TopicList should not be null", topicList);
            assertNotSame("TopicList should not empty.", 0,
                    topicList.mTopicsList.size());
        } catch (Exception e) {
            fail("Exception occurs when calling getProviderId() API. Exception is : "
                    + e.getLocalizedMessage());
        }
    }

    /**
     * @since 2016-09-30
     * @see none
     * @objective Test 'getProviderState()' function with positive basic way
     * @target ProviderState getProviderState() throws NSException
     * @test_data None
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback 7. Register Topic
     *                ("TestTopic")
     * @procedure Perform getProviderState() API
     * @post_condition None
     * @expected 1. No exception occurs 2. Should get ALLOW state
     */
    public void testGetProviderState_SRC_P() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            ProviderState providerState = provider.getProviderState();

            if (providerState == ProviderState.ALLOW
                    || providerState == ProviderState.DENY
                    || providerState == ProviderState.TOPIC) {
                Log.i(TAG, "Provider State may be ALLOW, DENY or TOPIC");
            } else {
                fail("ProviderState should be ALLOW, DENY or TOPIC state for subControllability = false");
            }
        } catch (Exception e) {
            fail("Exception occurs when calling getProviderState() API. Exception is : "
                    + e.getLocalizedMessage());
        }
    }

    /**
     * @since 2016-10-03
     * @see none
     * @objective Test 'subscribe()' function with positive basic way
     * @target void subscribe() throws NSException
     * @test_data None
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback
     * @procedure Perform subscribe() API
     * @post_condition None
     * @expected 1. No exception occurs 2. Should get ALLOW state
     */
    public void testSubscribe_SRC_P() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            provider.subscribe();
        } catch (Exception e) {
            fail("Exception occurs when calling subscribe() API. Exception is : "
                    + e.getLocalizedMessage());
        }
    }

    /**
     * @since 2016-10-03
     * @see none
     * @objective Test 'isSubscribed()' function with positive basic way
     * @target boolean isSubscribed() throws NSException
     * @test_data None
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback
     * @procedure Perform isSubscribed() without calling subscribe()
     * @post_condition None
     * @expected 1. No exception occurs 2. Should get false as return value
     */
    public void testIsSubscribe_SRC_P() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            assertEquals("Should not subscribed, before subscription", false,
                    provider.isSubscribed());
        } catch (Exception e) {
            fail("Exception occurs when calling subscribe() API. Exception is : "
                    + e.getLocalizedMessage());
        }
    }

    /**
     * @since 2016-10-03
     * @see none
     * @objective Test 'sendSyncInfo()' function with positive basic way
     * @target void sendSyncInfo(long messageId, SyncInfo.SyncType syncType)
     * @test_data None
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback
     * @procedure Perform sendSyncInfo() API
     * @post_condition None
     * @expected No exception occurs
     */
    public void testSendSyncInfo_SRC_P() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            provider.sendSyncInfo(mMessageId, SyncType.READ);
        } catch (Exception e) {
            fail("Exception occurs when calling sendSyncInfo() API. Exception is : "
                    + e.getLocalizedMessage());
        }
    }

    /**
     * @since 2016-10-03
     * @see none
     * @objective Test 'setListener()' function with positive basic way
     * @target void setListener(OnProviderStateListener onProviderStateListener,
     *         OnMessageReceivedListner onMessageReceivedListner,
     *         OnSyncInfoReceivedListner onSyncInfoReceivedListner)
     * @test_data OnProviderStateListener, OnMessageReceivedListner,
     *            OnSyncInfoReceivedListner
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback
     * @procedure Perform setListener() API
     * @post_condition None
     * @expected No exception occurs
     */
    public void testSetListener_SRC_P() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            provider.setListener(mHelper, mHelper, mHelper);
        } catch (Exception e) {
            fail("Exception occurs when calling setListener() API. Exception is : "
                    + e.getLocalizedMessage());
        }
    }

    /**
     * @since 2016-10-03
     * @see none
     * @objective Test 'setListener()' function with null
     *            OnProviderStateListener
     * @target void setListener(OnProviderStateListener onProviderStateListener,
     *         OnMessageReceivedListner onMessageReceivedListner,
     *         OnSyncInfoReceivedListner onSyncInfoReceivedListner)
     * @test_data OnProviderStateListener = null, OnMessageReceivedListner,
     *            OnSyncInfoReceivedListner
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback
     * @procedure Perform setListener() with null OnProviderStateListener
     * @post_condition None
     * @expected NSException should occur
     */
    public void testSetListenerWithStateListener_NV_N() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            provider.setListener(null, mHelper, mHelper);
            fail("Successfully set null as OnProviderStateListener");
        } catch (NSException nsE) {
        } catch (Exception e) {
            fail("Should throw NSException");
        }
    }

    /**
     * @since 2016-10-03
     * @see none
     * @objective Test 'setListener()' function with null
     *            OnMessageReceivedListner
     * @target void setListener(OnProviderStateListener onProviderStateListener,
     *         OnMessageReceivedListner onMessageReceivedListner,
     *         OnSyncInfoReceivedListner onSyncInfoReceivedListner)
     * @test_data OnProviderStateListener, OnMessageReceivedListner = null,
     *            OnSyncInfoReceivedListner
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback
     * @procedure Perform setListener() with null OnMessageReceivedListner
     * @post_condition None
     * @expected NSException should occur
     */
    public void testSetListenerWithMsgReceivedListener_NV_N() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            provider.setListener(mHelper, null, mHelper);
            fail("Successfully set null as OnMessageReceivedListner");
        } catch (NSException nsE) {
        } catch (Exception e) {
            fail("Should throw NSException");
        }
    }

    /**
     * @since 2016-10-03
     * @see none
     * @objective Test 'setListener()' function with null
     *            OnSyncInfoReceivedListner
     * @target void setListener(OnProviderStateListener onProviderStateListener,
     *         OnMessageReceivedListner onMessageReceivedListner,
     *         OnSyncInfoReceivedListner onSyncInfoReceivedListner)
     * @test_data OnProviderStateListener, OnMessageReceivedListner,
     *            OnSyncInfoReceivedListner = null
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback
     * @procedure Perform setListener() with null OnSyncInfoReceivedListner
     * @post_condition None
     * @expected NSException should occur
     */
    public void testSetListenerWithSyncInfoListener_NV_N() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            provider.setListener(mHelper, mHelper, null);
            fail("Successfully set null as OnSyncInfoReceivedListner");
        } catch (NSException nsE) {
        } catch (Exception e) {
            fail("Should throw NSException");
        }
    }

    /**
     * @since 2016-10-03
     * @see none
     * @objective Test 'updateTopicList()' function with positive basic way
     * @target int updateTopicList() throws NSException
     * @test_data None
     * @pre_condition 1. Get Provider Service Instance 2. Start Provider with
     *                consumer acceptor 3. Get Consumer Service Instance 4.
     *                Start Consumer Service 5. Wait for Provider Discovered
     *                Callback 6. Get Provider from callback 7. Register Topic
     *                ("TestTopic") 8. Get TopicList
     * @procedure Perform updateTopicList() API
     * @post_condition None
     * @expected 1. No exception occurs 2. Should return NS_OK (100)
     */
    public void testUpdateTopicList_SRC_P() {
        Provider provider = mHelper.getProvider(false);

        assertNotNull("Can't get any provider", provider);

        try {
            TopicsList topicList = provider.getTopicList();

            assertNotNull("TopicList should not be null", topicList);

            assertEquals("Should return NS_OK(100)", NS_OK,
                    provider.updateTopicList(topicList));
        } catch (Exception e) {
            fail("Exception occurs when calling getProviderId() API. Exception is : "
                    + e.getLocalizedMessage());
        }
    }
}
